### DISCUSSION TITLE
Reset camera of a model-viewer element when src is changed.

### DISCUSSION THREAD

**Discussion by noah-pm-dev** [google/model-viewer#4572] [Q&A]
I'm using `model-viewer` in my electron/react app, and I am using react's `useState()` to change the `src` value of the `model-viewer` element, thereby changing the model it is displaying.

I want to be able to reset the camera view to a certain position whenever the `src` attribute is updated. The issue is, the `camera-orbit` attribute doesn't actually change when you rotate the model. So the following code doesn't work to update the camera:
```js
    const [url, setUrl] = useState("../models/forks.glb");
    const [cameraOrbit, setCameraOrbit] = useState("45deg 55deg 4m")

    function changeModel(modelName: string) {
        setUrl(`../models/${modelName}.glb`)
        setCameraOrbit("45deg 55deg 4m")
    }
    
    return (
        <misc html>
        <ModelViewer modelPath={new URL(url, import.meta.url).href} cameraOrbit={cameraOrbit} />
        <misc html>
    )
```
`changeModel()` is run when certain buttons are clicked, and `setUrl()` works as intended.

`ModelViewer` is defined here:
```js
export default function ModelViewer(props: { modelPath: string, cameraOrbit: string }) {
    return ( <model-viewer src={props.modelPath} camera-orbit={props.cameraOrbit} camera-controls touch-action="pan-y" /> )
}
```

What would be a good way to do this?

**Comment by elalish** [comment]
I'm not familiar with React, but I assume there's a way to access our JS API in addition to our HTML API? Only the HTML attributes have the behavior you describe. If you instead select the MV element and `mv.cameraOrbit='auto auto auto';`, it will reset as desired.

**Reply by noah-pm-dev** [reply]
The following code somewhat works:
```js
    const [url, setUrl] = useState("../models/forks.glb");
    const defaultCameraOrbit = "45deg 55deg 5m";
    var mv = document.querySelector("model-viewer");

    function changeModel(modelName: string) {
        setUrl(`../models/${modelName}.glb`);
        setCameraOrbit(defaultCameraOrbit)
    }

    function setCameraOrbit(orbit: string) {
        mv!.cameraOrbit = orbit;
    }
```
However weirdly it only works if I run `changeModel` WITHOUT the url being different. If I run `changeModel` with an argument that will actually result in the model being changed, then the camera orbit does not reset. 

In the actual app this behavior manifests as the camera orbit only resetting if I click the button for the model that is already displaying. If I click a button that will change the model that is being displayed (changing the `src` attribute), then the camera orbit does not change.

So `setCameraOrbit` works fine on its own, but when run inside `changeModel` it only works if `setUrl` doesn't actually cause `url` to change. Probably a react thing so I will look more into how `useState` works.

**Comment by noah-pm-dev** [comment]
Because I was having some incredibly weird interactions between react and MV, I tested MV on its own, outside react, and the above solution did not work at all. Here is the code:
```html
<!DOCTYPE html>
<html>

<head>
    <script type="module" src="https://ajax.googleapis.com/ajax/libs/model-viewer/3.3.0/model-viewer.min.js"></script>
    <title>My Electron App</title>
</head>

<body>
    <model-viewer src=forks.glb camera-orbit="45deg 55deg 4m" camera-controls touch-action="none" interaction-prompt="none" />
    <button onclick="changeModel('table.glb')">Arm</button>
    <button onclick="changeModel('forks.glb')">Forks</button>
    <script>
        function changeModel(model) {
            var mv = document.querySelector("model-viewer")
            mv.setAttribute("src", model)
            mv.cameraOrbit = "45deg 55deg 5m"
        }
    </script>
</body>

</html>
```
Setting the `src` attribute does change the model, but `cameraOrbit` does not seem to do anything at all, whereas when using react it at least worked when there wasn't any severe react magic going on.

**Reply by elalish** [reply]
I believe you're running into async difficulties. After you set `src`, you need to await our "load" event before you set the camera, since "auto" bases the camera on the model's dimensions, which it doesn't know until it's been loaded.

**Reply by noah-pm-dev** [reply]
I'm not actually using `auto` in my code though, I am specifying `"45deg 55deg 5m"`.

**Reply by elalish** [reply]
Yes, but it's an option, so that's how our API is designed. Plus we're a Lit element, which means basically everything is very async, for performance.

**Reply by noah-pm-dev** [reply]
Oh sorry, I thought you were saying that it only awaited if it was set to auto. That makes sense.
